Kernel/Devices/AHCIPort.cpp

```diff
@@ -0,0 +1,508 @@
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <AK/Memory.h>
+#include <Kernel/Devices/AHCIController.h>
+#include <Kernel/Devices/AHCIPort.h>
+#include <Kernel/VM/MemoryManager.h>
+
+#define FIS_TYPE_REG_H2D 0x27
+#define FIS_TYPE_PIO_SETUP 0x5F
+
+#define DEV_CLB 0x00
+#define DEV_CLBU 0x04
+#define DEV_FB 0x08
+#define DEV_FBU 0x0C
+#define DEV_IS 0x10
+#define DEV_IE 0x14
+#define DEV_CMD 0x18
+#define DEV_TFD 0x20
+#define DEV_SIG 0x24
+#define DEV_SSTS 0x28
+#define DEV_SERR 0x30
+#define DEV_SACT 0x34
+#define DEV_CI 0x38
+
+#define CMD_PRDTL 0x02
+#define CMD_PRDBC 0x04
+#define CMD_CTBA 0x08
+#define CMD_CTBAU 0x0C
+
+#define IRQ_TFES 0x1 << 30
+#define IRQ_HBFS 0x1 << 29
+#define IRQ_HBDS 0x1 << 28
+#define IRQ_IFS 0x1 << 27
+#define IRQ_INFS 0x1 << 26
+#define IRQ_OFS 0x1 << 24
+
+#define SATA_SIG_ATA 0x00000101
+#define SATA_SIG_ATAPI 0xEB140101
+#define SATA_SIG_SEMB 0xC33C0101
+#define SATA_SIG_PM 0x96690101
+
+#define HBA_PORT_IPM_ACTIVE 1
+#define HBA_PORT_DET_PRESENT 3
+
+#define ATA_DEV_BUSY 0x80
+#define ATA_DEV_DRQ 0x08
+#define ATA_CMD_READ_DMA_EX 0x25
+#define ATA_CMD_IDENTIFY 0xEC
+
+#define HBA_PxCMD_ST 0x0001
+#define HBA_PxCMD_FRE 0x0010
+#define HBA_PxCMD_FR 0x4000
+#define HBA_PxCMD_CR 0x8000
+#define HBA_PxIS_TFES 1 << 30
+
+#define SLOT_CFL 0x1F
+#define SLOT_ATAPI 0x1 << 5
+#define SLOT_WRITE 0x1 << 6
+#define SLOT_PREFETCH 0x1 << 7
+
+#define SLOT_RESET 0x1 << 8
+#define SLOT_BIST 0x1 << 9
+#define SLOT_CLEAR 0x1 << 10
+#define SLOT_PORT_MULT 0xF000
+#define DATA_ENTRY_DBC 0x3fffff
+#define DATA_ENTRY_I 0x1 << 31
+#define FIS_H2D_PMPORT 0xF
+#define FIS_H2D_MODE 0x1 << 7
+
+#define LSB(x) ((x)&0xFF)
+#define MSB(x) (((x) >> 8) & 0xFF)
+
+// Only store at least one page inside the buffer
+#define SECTOR_SIZE 512
```

Comment:
I'd call this `ATA_LOGICAL_SECTOR_SIZE` instead, since it represents the common sector size of ATA-compatible harddrives today. Many new harddrives internally use 4k sized sectors, and optical discs can have sectors of 2048 bytes.